### PR TITLE
MULE-16141 : SDK is not validating input POJOs to have a default constructor

### PR DESCRIPTION
**Pull request by JuanAller**

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 1]
```diff
@@ -132,6 +146,17 @@ private void checkInvalidFieldAnnotations(NamedObject model, String typeName, Fi
...
       }
+
+      private boolean isPojoWithDefaultConstructor(Class<?> clazz) {
+        return (!clazz.isInterface() && !isAbstract(clazz.getModifiers())
+            && Stream.of(clazz.getConstructors())
+                .noneMatch(c -> c.getParameterCount() == 0));
+      }
+
+      private Class<?> getTypedValueGenericClass(Type type) {
+        return type.getGenerics().size() != 0 ? type.getGenerics().get(0).getConcreteType().getDeclaringClass().get()
```
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Code Smell Code Smell: Use isEmpty() to check whether the collection is empty or not. ([squid:S1155](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1155))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhiR_G7otIXnmsUW4nN&open=AWhiR_G7otIXnmsUW4nN)

[ISSUE_KEY]: AWhiR_G7otIXnmsUW4nN

Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhiR_G7otIXnmsUW4nM&open=AWhiR_G7otIXnmsUW4nM)

[ISSUE_KEY]: AWhiR_G7otIXnmsUW4nM

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 2]
```diff
@@ -132,6 +146,17 @@ private void checkInvalidFieldAnnotations(NamedObject model, String typeName, Fi
           }
         }
       }
+
+      private boolean isPojoWithDefaultConstructor(Class<?> clazz) {
+        return (!clazz.isInterface() && !isAbstract(clazz.getModifiers())
+            && Stream.of(clazz.getConstructors())
+                .noneMatch(c -> c.getParameterCount() == 0));
```
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Code Smell Code Smell: Remove the parentheses around the "c" parameter ([squid:S1611](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1611))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhdy-0YotIXnmsUGksQ&open=AWhdy-0YotIXnmsUGksQ)

[ISSUE_KEY]: AWhdy-0YotIXnmsUGksQ

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java (hunk 1 above)
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Code Smell Code Smell: Use isEmpty() to check whether the collection is empty or not. ([squid:S1155](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1155))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhiR_G7otIXnmsUW4nN&open=AWhiR_G7otIXnmsUW4nN)

[ISSUE_KEY]: AWhiR_G7otIXnmsUW4nN

Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhiR_G7otIXnmsUW4nM&open=AWhiR_G7otIXnmsUW4nM)

[ISSUE_KEY]: AWhiR_G7otIXnmsUW4nM

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 3]
```diff
@@ -132,6 +147,18 @@ private void checkInvalidFieldAnnotations(NamedObject model, String typeName, Fi
...
+
+      private boolean isPojoWithoutDefaultConstructor(Class<?> clazz) {
+        return (!clazz.isInterface() && !isAbstract(clazz.getModifiers())
+            && Stream.of(clazz.getConstructors())
+                .noneMatch(c -> c.getParameterCount() == 0));
+      }
+
+      private Class<?> getTypedValueGenericClass(Type type) {
+        return !CollectionUtils.isEmpty(type.getGenerics())
+            ? type.getGenerics().get(0).getConcreteType().getDeclaringClass().get()
```
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhx-i7aotIXnmsUtVBm&open=AWhx-i7aotIXnmsUtVBm)

[ISSUE_KEY]: AWhx-i7aotIXnmsUtVBm

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 4]
```diff
@@ -132,6 +148,18 @@ private void checkInvalidFieldAnnotations(NamedObject model, String typeName, Fi
...
+
+      private boolean isPojoWithoutDefaultConstructor(Class<?> clazz) {
+        return (!clazz.isInterface() && !isAbstract(clazz.getModifiers())
+            && Stream.of(clazz.getConstructors())
+                .noneMatch(c -> c.getParameterCount() == 0));
+      }
+
+      private Class<?> getTypedValueGenericClass(Type type) {
+        return !CollectionUtils.isEmpty(type.getGenerics())
+            ? type.getGenerics().stream().findFirst().get().getConcreteType().getDeclaringClass().get()
```
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhykiEpotIXnmsUzH_b&open=AWhykiEpotIXnmsUzH_b)

[ISSUE_KEY]: AWhykiEpotIXnmsUzH_b

Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Bug Bug: Call "Optional#isPresent()" before accessing the value. ([squid:S3655](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS3655))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhykiEpotIXnmsUzH_c&open=AWhykiEpotIXnmsUzH_c)

[ISSUE_KEY]: AWhykiEpotIXnmsUzH_c

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 5]
```diff
@@ -32,12 +33,17 @@
...
 import org.mule.runtime.module.extension.api.loader.java.type.FieldElement;
+import org.mule.runtime.module.extension.api.loader.java.type.Type;
 import org.mule.runtime.module.extension.internal.loader.java.type.property.ExtensionParameterDescriptorModelProperty;
+import org.springframework.util.CollectionUtils;
 
 import java.lang.annotation.Annotation;
+import static java.lang.reflect.Modifier.isAbstract;
 import java.util.HashSet;
 import java.util.Objects;
+import java.util.Optional;
```
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: Code Smell Code Smell: Remove this unused import 'java.util.Optional'. ([squid:UselessImportCheck](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AUselessImportCheck))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWhykiEpotIXnmsUzH_d&open=AWhykiEpotIXnmsUzH_d)

[ISSUE_KEY]: AWhykiEpotIXnmsUzH_d

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 6]
```diff
@@ -102,6 +108,16 @@ public void visitObject(ObjectType objectType) {
               .map(descriptor -> descriptor.getExtensionParameter().getType())
               .ifPresent(type -> {
                 final String typeName = type.getName();
+
+                Class<?> clazz = type.getDeclaringClass().get().equals(TypedValue.class)
+                    ? getTypedValueGenericClass(type)
+                    : type.getDeclaringClass().get();
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: How are you sure that type.getDeclaringClass() is not empty() ?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java [hunk 7]
```diff
@@ -132,6 +148,20 @@ private void checkInvalidFieldAnnotations(NamedObject model, String typeName, Fi
...
       }
+
+      private boolean isPojoWithoutDefaultConstructor(Class<?> clazz) {
+        return (!clazz.isInterface() && !isAbstract(clazz.getModifiers())
+            && Stream.of(clazz.getConstructors())
+                .noneMatch(c -> c.getParameterCount() == 0));
+      }
+
+      private Class<?> getTypedValueGenericClass(Type type) {
+        Optional<Class<?>> clazz = type.getGenerics().stream().findFirst()
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/ParameterTypeModelValidator.java]: I think this method name does not reflect what it actually does. If the type get generics it gets the first one. If not, it returns Object.class.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    });
+  }
+
+  private Optional<Type> getFirstGenericType(Type typeWithGenerics) {
+    return !CollectionUtils.isEmpty(typeWithGenerics.getGenerics())
+        ? Optional.ofNullable(typeWithGenerics.getGenerics().get(0).getConcreteType()) : Optional.empty();
+  }
+
+  @FunctionalInterface
+  private interface TypeValidator<M, T, PR> {
```
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: Code Smell Code Smell: Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. ([squid:S00119](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS00119))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWjKagnwotIXnmsUX912&open=AWjKagnwotIXnmsUX912)

[ISSUE_KEY]: AWjKagnwotIXnmsUX912

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+      }
+
+      @Override
+      protected void onConnectionProvider(HasConnectionProviderModels owner, ConnectionProviderModel model) {
+        validateOverridesEqualsAndHashCode(model, problemsReporter);
+      }
+
+      @Override
+      protected void onParameter(ParameterizedModel owner, ParameterGroupModel groupModel, ParameterModel model) {
+        validatePojoHasDefaultConstructor(owner, problemsReporter);
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: Here you are running the validation for each parameter, as many times as parameters its owner has. You should only make the validation for the model here.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+import org.mule.runtime.module.extension.api.loader.java.type.MethodElement;
+import org.mule.runtime.module.extension.api.loader.java.type.Type;
+import org.mule.runtime.module.extension.internal.loader.java.type.property.ExtensionParameterDescriptorModelProperty;
+
+import java.util.Optional;
+
+import org.apache.commons.collections.CollectionUtils;
+
+/**
+ * Validates that POJOs are valid for the model
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: Briefly describe what it needs to be valid

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+
+      @Override
+      protected void onParameter(ParameterizedModel owner, ParameterGroupModel groupModel, ParameterModel model) {
+        validatePojoHasDefaultConstructor(owner, problemsReporter);
+      }
+    }.walk(extensionModel);
+  }
+
+  private void iterateParametersAndValidate(ParameterizedModel model, ProblemsReporter problemsReporter,
+                                            TypeValidator<ParameterizedModel, Type, ProblemsReporter> validator) {
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: TypeValidator<ParameterModel, Type, ProblemsReporter>, you should validate a parameter, not its owner.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+  private void validateOverridesEqualsAndHashCode(ParameterizedModel model, ProblemsReporter reporter) {
+    iterateParametersAndValidate(model, reporter, (parameterizedModel, type, problemReporter) -> {
+      ClassInformationAnnotation classInformationAnnotation = type.getClassInformation();
+      if (!classInformationAnnotation.isInterface() && !classInformationAnnotation.isAbstract()
+          && type.asMetadataType() instanceof ObjectType) {
+        Optional<MethodElement> equals = type.getMethod("equals", Object.class);
+        Optional<MethodElement> hashCode = type.getMethod("hashCode");
+        if (!equals.isPresent() || !hashCode.isPresent()) {
+          reporter
+              .addError(new Problem(model,
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: The problem should be pointed to the parameter, not its owner

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+          parameterModel.getModelProperty(ExtensionParameterDescriptorModelProperty.class);
+      if (modelProperty.isPresent()) {
+        Type type = modelProperty.get().getExtensionParameter().getType();
+        validator.validateType(model, type, problemsReporter);
+      }
+    });
+  }
+
+  private void validateOverridesEqualsAndHashCode(ParameterizedModel model, ProblemsReporter reporter) {
+    iterateParametersAndValidate(model, reporter, (parameterizedModel, type, problemReporter) -> {
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: the lambda function should receive a parameterModel

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+          reporter
+              .addError(new Problem(model,
+                                    format("Type '%s' must override equals and hashCode",
+                                           type.getName())));
+        }
+      }
+    });
+  }
+
+  private void validatePojoHasDefaultConstructor(ParameterizedModel owner, ProblemsReporter problemsReporter) {
```
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: This one should directly receive the parameter model and validate it

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java (hunk 8 above)
Comment (ndinu) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: Use single letters for java generic

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidatorTestCase.java [hunk 15]
```diff
@@ -61,6 +61,14 @@ public void pojoImplementsEqualsAndHashCode() {
...
   }
 
+  @Test
+  public void pojoConfigurationMustHaveDefaultConstructor() {
+    expectedException.expect(IllegalModelDefinitionException.class);
+    expectedException.expectMessage("Type 'InvalidPojoRequiresDefaultConstructor' does not have a default constructor");
+    validate(InvalidConfigParameterTestConnector.class);
+  }
+
+
```
Comment (ndinu) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidatorTestCase.java]: Add a failing test that checks the amount of errors with an extension that has a config with 2 pojos, that only one has a default constructor. Amount of errors should be 1.

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java (hunk 8 above)
Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: Code Smell Code Smell: Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. ([squid:S00119](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS00119))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWjKagnwotIXnmsUX912&open=AWjKagnwotIXnmsUX912)

[ISSUE_KEY]: AWjKagnwotIXnmsUX912

Comment (sonar-mulesoft) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/validation/PojosModelValidator.java]: Code Smell Code Smell: Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. ([squid:S00119](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS00119))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7615&issues=AWjKagnwotIXnmsUX912&open=AWjKagnwotIXnmsUX912)

[ISSUE_KEY]: AWjKagnwotIXnmsUX912

**Comment by JuanAller** [comment]
--validate

**Pull request merged**